File: app/src/main/java/miles/diary/ui/activity/EntryActivity.java
package miles.diary.ui.activity;

import android.annotation.SuppressLint;
import android.content.Context;
import android.content.Intent;
import android.graphics.Bitmap;
import android.graphics.Color;
import android.graphics.PorterDuff;
import android.net.Uri;
import android.os.Bundle;
import android.support.design.widget.FloatingActionButton;
import android.support.design.widget.Snackbar;
import android.support.v4.content.ContextCompat;
import android.support.v7.graphics.Palette;
import android.text.SpannableString;
import android.text.Spanned;
import android.text.style.RelativeSizeSpan;
import android.transition.ArcMotion;
import android.transition.ChangeImageTransform;
import android.transition.Transition;
import android.transition.TransitionSet;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.view.animation.DecelerateInterpolator;
import android.widget.TextView;
import android.widget.Toolbar;

import com.bumptech.glide.Glide;
import com.bumptech.glide.request.RequestListener;
import com.bumptech.glide.request.target.Target;
import com.google.gson.Gson;

import javax.inject.Inject;

import butterknife.Bind;
import butterknife.OnClick;
import icepick.State;
import miles.diary.DiaryApplication;
import miles.diary.R;
import miles.diary.data.api.Repository;
import miles.diary.data.model.realm.Entry;
import miles.diary.data.model.weather.WeatherResponse;
import miles.diary.data.rx.ActivitySubscriber;
import miles.diary.data.rx.DataTransaction;
import miles.diary.ui.PaletteWindows;
import miles.diary.ui.TypefacerSpan;
import miles.diary.ui.transition.RoundedImageViewTransition;
import miles.diary.ui.transition.SimpleTransitionListener;
import miles.diary.ui.widget.RoundedImageView;
import miles.diary.ui.widget.TypefaceIconTextView;
import miles.diary.ui.widget.TypefaceTextView;
import miles.diary.util.AnimUtils;
import miles.diary.util.Logg;
import miles.diary.util.TextUtils;
import miles.diary.util.ViewUtils;

/**
 * Created by mbpeele on 2/8/16.
 */
public class EntryActivity extends TransitionActivity {

    public final static String INTENT_KEY = "data";
    public final static String INTENT_ACTION = "action";
    public final static int REQUEST_EDIT_ENTRY = 1;

    public enum Action {
        EDIT,
        DELETE
    }

    @Inject
    Repository repository;

    @Bind(R.id.activity_entry_place_photos)
    FloatingActionButton photosFab;
    @Bind(R.id.activity_entry_toolbar)
    Toolbar toolbar;
    @Bind(R.id.activity_entry_body)
    TypefaceTextView body;
    @Bind(R.id.activity_entry_image)
    RoundedImageView image;
    @Bind(R.id.activity_entry_place)
    TypefaceTextView place;
    @Bind(R.id.activity_entry_date)
    TypefaceTextView date;
    @Bind(R.id.activity_entry_weather)
    TypefaceIconTextView weather;

    private Entry entry;
    @State boolean dataChanged;

    public static Intent newIntent(Context context, Entry entry) {
        Intent intent = new Intent(context, EntryActivity.class);
        intent.putExtra(EntryActivity.INTENT_KEY, entry.getDateMillis());
        return intent;
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_entry);

        setActionBar(toolbar);
        if (getActionBar() != null) {
            getActionBar().setDisplayHomeAsUpEnabled(true);
        }

        entry = repository.get(Entry.class, getIntent().getLongExtra(INTENT_KEY, -1));
        updateView(entry);
    }

    @Override
    public void inject(DiaryApplication diaryApplication) {
        diaryApplication.getContextComponent().inject(this);
    }

    @Override
    public void onBackPressed() {
        if (dataChanged) {
            Logg.log("DATA CHANGED");
            setResultAction(Action.EDIT);
        } else {
            super.onBackPressed();
        }
    }

    @Override
    boolean overrideTransitions() {
        return false;
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.menu_entry, menu);
        for (int i = 0; i < menu.size(); i++) {
            MenuItem item = menu.getItem(i);
            if (entry.getUri() == null) {
                item.getIcon().mutate().setColorFilter(Color.BLACK, PorterDuff.Mode.SRC_IN);
            }
        }
        return super.onCreateOptionsMenu(menu);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.menu_entry_edit:
                Intent intent = new Intent(this, NewEntryActivity.class);
                intent.putExtra(EntryActivity.INTENT_KEY, entry.getDateMillis());
                startActivityForResult(intent, REQUEST_EDIT_ENTRY);
                break;
            case R.id.menu_entry_delete:
                setResultAction(Action.DELETE);
                break;
            case android.R.id.home:
                finishAfterTransition();
                break;
        }
        return super.onOptionsItemSelected(item);
    }

    @Override
    @SuppressWarnings("ConstantConditions")
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        switch (requestCode) {
            case REQUEST_EDIT_ENTRY:
                if (resultCode == RESULT_OK) {
                    Bundle bundle = data.getExtras();
                    final String body = bundle.getString(NewEntryActivity.BODY);
                    final Uri uri = bundle.getParcelable(NewEntryActivity.URI);
                    final String placeName = bundle.getString(NewEntryActivity.PLACE_NAME);
                    final String placeId = bundle.getString(NewEntryActivity.PLACE_ID);

                    repository.updateObject(new DataTransaction<Entry>() {
                        @Override
                        public Entry call() {
                            return entry.update(body, uri, placeName, placeId);
                        }
                    }).subscribe(new ActivitySubscriber<Entry>(this) {
                        @Override
                        public void onNext(Entry entry1) {
                            dataChanged = true;
                            entry = entry1;
                            updateView(entry);
                        }
                    });
                }
                break;
            default:
                super.onActivityResult(requestCode, resultCode, data);
        }
    }

    @Override
    void onEnter(final ViewGroup root, Intent calledIntent, boolean hasSavedInstanceState) {
        if (entry.getUri() == null) {
            int color = ContextCompat.getColor(this, R.color.dark_icons);
            toolbar.getNavigationIcon().mutate().setColorFilter(color, PorterDuff.Mode.SRC_IN);
            toolbar.setTitleTextColor(color);
        } else {
            ArcMotion arcMotion = new ArcMotion();
            arcMotion.setMinimumHorizontalAngle(50f);
            arcMotion.setMinimumVerticalAngle(50f);

            RoundedImageViewTransition reveal = new RoundedImageViewTransition(
                    Math.max(image.getWidth(), image.getHeight()) / 2f, 0);
            reveal.addTarget(image);
            reveal.setPathMotion(arcMotion);

            reveal.addListener(new SimpleTransitionListener() {
                @Override
                public void onTransitionStart(Transition transition) {
                    toolbar.setVisibility(View.GONE);
                    slideUpView(root, AnimUtils.shortAnim(EntryActivity.this));
                }

                @Override
                public void onTransitionEnd(Transition transition) {
                    AnimUtils.visible(toolbar).start();
                    ViewUtils.setZoomControls(image);
                }
            });

            TransitionSet returnSet = new TransitionSet();

            RoundedImageViewTransition unreveal = new RoundedImageViewTransition(
                    0, Math.min(image.getWidth(), image.getHeight()) / 2f);
            unreveal.addTarget(image);
            unreveal.setPathMotion(arcMotion);
            unreveal.addListener(new SimpleTransitionListener() {
                @Override
                public void onTransitionStart(Transition transition) {
                    toolbar.setVisibility(View.GONE);
                }
            });

            returnSet.addTransition(unreveal);
            returnSet.addTransition(new ChangeImageTransform());

            getWindow().setSharedElementEnterTransition(reveal);
            getWindow().setSharedElementReturnTransition(returnSet);
        }
    }

    @Override
    void onExit(ViewGroup root) {
    }

    @SuppressLint("SetTextI18n")
    private void updateView(Entry entry) {
        ViewUtils.mutate(place, place.getCurrentTextColor());
        ViewUtils.mutate(date, place.getCurrentTextColor());

        String text = "Dear Diary, " +
                TextUtils.repeat(2, TextUtils.LINE_SEPERATOR) +
                TextUtils.repeat(5, TextUtils.TAB) +
                entry.getBody();

        SpannableString spannableString = new SpannableString(text);
        spannableString.setSpan(new RelativeSizeSpan(1.4f), 0, 12, Spanned.SPAN_INCLUSIVE_INCLUSIVE);
        spannableString.setSpan(new TypefacerSpan(TextUtils.getFont(this, getString(R.string.default_font))),
                12, spannableString.length(), Spanned.SPAN_INCLUSIVE_INCLUSIVE);
        body.setText(spannableString, TextView.BufferType.SPANNABLE);

        String placeName = entry.getPlaceName();
        if (placeName != null) {
            place.setVisibility(View.VISIBLE);
            place.setText(placeName);
        } else {
            photosFab.setVisibility(View.GONE);
            place.setVisibility(View.GONE);
        }

        date.setText(TextUtils.formatDate(entry.getDate()) + TextUtils.LINE_SEPERATOR +
                TextUtils.formatTime(entry.getDate()));

        String string = entry.getWeather();
        if (string != null) {
            weather.setVisibility(View.VISIBLE);
            WeatherResponse weatherResponse = new Gson().fromJson(string, WeatherResponse.class);
            weather.setText(weatherResponse.getOneLineTemperatureString());
        } else {
            weather.setVisibility(View.GONE);
        }

        if (entry.getUri() != null) {
            postponeEnterTransition();

            image.setVisibility(View.VISIBLE);

            Glide.with(this)
                    .fromString()
                    .asBitmap()
                    .load(entry.getUri())
                    .centerCrop()
                    .listener(new RequestListener<String, Bitmap>() {
                        @Override
                        public boolean onException(Exception e, String model, Target<Bitmap> target, boolean isFirstResource) {
                            Logg.log(e);
                            return false;
                        }

                        @Override
                        public boolean onResourceReady(final Bitmap resource, String model, Target<Bitmap> target,
                                                       boolean isFromMemoryCache, boolean isFirstResource) {
                            Palette.from(resource)
                                    .maximumColorCount(3)
                                    .clearFilters()
                                    .generate(new PaletteWindows(EntryActivity.this, resource));

                            startPostponedEnterTransition();
                            return false;
                        }
                    })
                    .into(image);
        } else {
            image.setVisibility(View.GONE);
        }
    }

    private void slideUpView(ViewGroup root, int duration) {
        ViewGroup linearLayout = (ViewGroup) ((ViewGroup) root.getChildAt(1)).getChildAt(0);
        final float offset = linearLayout.getHeight() / 2f;
        for (int i = 0; i < linearLayout.getChildCount(); i++) {
            View v = linearLayout.getChildAt(i);

            v.setTranslationY(offset);
            v.setAlpha(0f);

            v.animate()
                    .translationY(0f)
                    .alpha(1f)
                    .setDuration(duration)
                    .setInterpolator(new DecelerateInterpolator())
                    .setStartDelay(150 + 50 * i)
                    .start();
        }
    }

    private void setResultAction(Action action) {
        Intent intent = new Intent();
        intent.putExtra(INTENT_KEY, action);
        intent.putExtra(INTENT_ACTION, entry.getDateMillis());
        setResult(RESULT_OK, intent);
        finish();
    }

    @OnClick(R.id.activity_entry_place_photos)
    protected void fabClick() {
        if (hasConnection()) {
            Intent intent = new Intent(this, PlacePhotosActivity.class);
            intent.putExtra(PlacePhotosActivity.ID, entry.getPlaceId());
            intent.putExtra(PlacePhotosActivity.NAME, entry.getPlaceName());
            startActivity(intent);
        } else {
            Snackbar.make(root, R.string.error_no_internet, Snackbar.LENGTH_SHORT).show();
        }
    }
}

File: app/src/main/java/miles/diary/ui/activity/EntryActivity.kt
package miles.diary.ui.activity

import android.annotation.SuppressLint
import android.app.Activity
import android.content.Context
import android.content.Intent
import android.graphics.Color
import android.graphics.PorterDuff
import android.net.Uri
import android.os.Bundle
import android.support.design.widget.FloatingActionButton
import android.support.design.widget.Snackbar
import android.support.v4.content.ContextCompat
import android.text.SpannableString
import android.text.Spanned
import android.text.style.RelativeSizeSpan
import android.transition.ArcMotion
import android.transition.ChangeImageTransform
import android.transition.Transition
import android.transition.TransitionSet
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import android.view.animation.DecelerateInterpolator
import android.widget.TextView
import android.widget.Toolbar
import butterknife.BindView
import butterknife.OnClick
import com.google.gson.Gson
import miles.diary.R
import miles.diary.data.model.realm.Entry
import miles.diary.data.model.weather.WeatherResponse
import miles.diary.data.rx.ActivitySubscriber
import miles.diary.ui.TypefacerSpan
import miles.diary.ui.transition.RoundedImageViewTransition
import miles.diary.ui.transition.SimpleTransitionListener
import miles.diary.ui.widget.RoundedImageView
import miles.diary.ui.widget.TypefaceIconTextView
import miles.diary.ui.widget.TypefaceTextView
import miles.diary.util.AnimUtils
import miles.diary.util.TextUtils
import miles.diary.util.ViewUtils

/**
 * Created by mbpeele on 2/8/16.
 */
class EntryActivity : TransitionActivity() {

    @BindView(R.id.activity_entry_place_photos)
    internal lateinit var photosFab: FloatingActionButton
    @BindView(R.id.activity_entry_toolbar)
    internal lateinit var toolbar: Toolbar
    @BindView(R.id.activity_entry_body)
    internal lateinit var body: TypefaceTextView
    @BindView(R.id.activity_entry_image)
    internal lateinit var image: RoundedImageView
    @BindView(R.id.activity_entry_place)
    internal lateinit var place: TypefaceTextView
    @BindView(R.id.activity_entry_date)
    internal lateinit var date: TypefaceTextView
    @BindView(R.id.activity_entry_weather)
    internal lateinit var weatherView: TypefaceIconTextView

    private lateinit var entry: Entry

    enum class Action {
        EDIT,
        DELETE
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_entry)

        toolbar.title = ""
        setActionBar(toolbar)
        if (actionBar != null) {
            actionBar.setDisplayHomeAsUpEnabled(true)
        }

        entry = repository.get(Entry::class.java, intent.getLongExtra(INTENT_KEY, -1))
        updateView(entry)
    }

    override fun overrideTransitions(): Boolean {
        return false
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.menu_entry, menu)
        (0 until menu.size())
                .map { menu.getItem(it) }
                .forEach {
                    if (entry.uri == null) {
                        it.icon.mutate().setColorFilter(Color.BLACK, PorterDuff.Mode.SRC_IN)
                    }
                }
        return super.onCreateOptionsMenu(menu)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.menu_entry_edit -> {
                val intent = Intent(this, NewEntryActivity::class.java)
                intent.putExtra(EntryActivity.INTENT_KEY, entry.dateMillis)
                startActivityForResult(intent, REQUEST_EDIT_ENTRY)
            }
            R.id.menu_entry_delete -> setResultAction(Action.DELETE)
            android.R.id.home -> finishAfterTransition()
        }
        return super.onOptionsItemSelected(item)
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent) {
        when (requestCode) {
            REQUEST_EDIT_ENTRY -> if (resultCode == Activity.RESULT_OK) {
                val bundle = data.extras
                val body = bundle.getString(NewEntryActivity.BODY)
                val uri = bundle.getParcelable<Uri>(NewEntryActivity.URI)
                val placeName = bundle.getString(NewEntryActivity.PLACE_NAME)
                val placeId = bundle.getString(NewEntryActivity.PLACE_ID)

                repository.updateObject({ entry.update(body, uri, placeName, placeId) }).subscribe(object : ActivitySubscriber<Entry>(this) {
                    override fun onNext(entry1: Entry) {
                        entry = entry1
                        updateView(entry)
                    }
                })
            }
            else -> super.onActivityResult(requestCode, resultCode, data)
        }
    }

    override fun onEnter(root: ViewGroup, calledIntent: Intent, hasSavedInstanceState: Boolean) {
        if (entry.uri == null) {
            val color = ContextCompat.getColor(this, R.color.dark_icons)
            toolbar.navigationIcon.mutate().setColorFilter(color, PorterDuff.Mode.SRC_IN)
            toolbar.setTitleTextColor(color)
        } else {
            val arcMotion = ArcMotion()
            arcMotion.minimumHorizontalAngle = 50f
            arcMotion.minimumVerticalAngle = 50f

            val reveal = RoundedImageViewTransition(
                    Math.max(image.width, image.height) / 2f, 0f)
            reveal.addTarget(image)
            reveal.pathMotion = arcMotion

            reveal.addListener(object : SimpleTransitionListener() {
                override fun onTransitionStart(transition: Transition) {
                    toolbar.visibility = View.GONE
                    slideUpView(root, AnimUtils.shortAnim(this@EntryActivity))
                }

                override fun onTransitionEnd(transition: Transition) {
                    AnimUtils.visible(toolbar).start()
                    ViewUtils.setZoomControls(image)
                }
            })

            val returnSet = TransitionSet()

            val unreveal = RoundedImageViewTransition(
                    0f, Math.min(image.width, image.height) / 2f)
            unreveal.addTarget(image)
            unreveal.pathMotion = arcMotion
            unreveal.addListener(object : SimpleTransitionListener() {
                override fun onTransitionStart(transition: Transition) {
                    toolbar.visibility = View.GONE
                }
            })

            returnSet.addTransition(unreveal)
            returnSet.addTransition(ChangeImageTransform())

            window.sharedElementEnterTransition = reveal
            window.sharedElementReturnTransition = returnSet
        }
    }

    internal override fun onExit(root: ViewGroup) {}

    @SuppressLint("SetTextI18n")
    private fun updateView(entry: Entry) {
        ViewUtils.mutate(place, place.currentTextColor)
        ViewUtils.mutate(date, place.currentTextColor)

        val text = "Dear Diary, " +
                TextUtils.repeat(2, TextUtils.LINE_SEPERATOR) +
                TextUtils.repeat(5, TextUtils.TAB) +
                entry.body

        val spannableString = SpannableString(text)
        spannableString.setSpan(RelativeSizeSpan(1.4f), 0, 12, Spanned.SPAN_INCLUSIVE_INCLUSIVE)
        spannableString.setSpan(TypefacerSpan(TextUtils.getFont(this, getString(R.string.default_font))),
                12, spannableString.length, Spanned.SPAN_INCLUSIVE_INCLUSIVE)
        body.setText(spannableString, TextView.BufferType.SPANNABLE)

        val placeName = entry.placeName
        if (placeName != null) {
            place.visibility = View.VISIBLE
            place.text = placeName
        } else {
            photosFab.visibility = View.GONE
            place.visibility = View.GONE
        }

        date.text = TextUtils.formatDate(entry.date) + TextUtils.LINE_SEPERATOR +
                TextUtils.formatTime(entry.date)

        val string = entry.weather
        if (string != null) {
            weatherView.visibility = View.VISIBLE
            val weatherResponse = Gson().fromJson(string, WeatherResponse::class.java)
            weatherView.text = weatherResponse.oneLineTemperatureString
        } else {
            weatherView.visibility = View.GONE
        }

        if (entry.uri != null) {
            postponeEnterTransition()

            image.visibility = View.VISIBLE

//            Glide.with(this)
//                    .fromString()
//                    .asBitmap()
//                    .load(entry.uri)
//                    .centerCrop()
//                    .listener(object : RequestListener<String, Bitmap> {
//                        fun onException(e: Exception, model: String, target: Target<Bitmap>, isFirstResource: Boolean): Boolean {
//                            Logg.log(e)
//                            return false
//                        }
//
//                        fun onResourceReady(resource: Bitmap, model: String, target: Target<Bitmap>,
//                                            isFromMemoryCache: Boolean, isFirstResource: Boolean): Boolean {
//                            Palette.from(resource)
//                                    .maximumColorCount(3)
//                                    .clearFilters()
//                                    .generate(PaletteWindows(this@EntryActivity, resource))
//
//                            startPostponedEnterTransition()
//                            return false
//                        }
//                    })
//                    .into(image)
        } else {
            image.visibility = View.GONE
        }
    }

    private fun slideUpView(root: ViewGroup, duration: Int) {
        val linearLayout = (root.getChildAt(1) as ViewGroup).getChildAt(0) as ViewGroup
        val offset = linearLayout.height / 2f
        for (i in 0 until linearLayout.childCount) {
            val v = linearLayout.getChildAt(i)

            v.translationY = offset
            v.alpha = 0f

            v.animate()
                    .translationY(0f)
                    .alpha(1f)
                    .setDuration(duration.toLong())
                    .setInterpolator(DecelerateInterpolator())
                    .setStartDelay((150 + 50 * i).toLong())
                    .start()
        }
    }

    private fun setResultAction(action: Action) {
        val intent = Intent()
        intent.putExtra(INTENT_KEY, action)
        intent.putExtra(INTENT_ACTION, entry.dateMillis)
        setResult(Activity.RESULT_OK, intent)
        finish()
    }

    @OnClick(R.id.activity_entry_place_photos)
    protected fun fabClick() {
        if (hasConnection()) {
            val intent = Intent(this, PlacePhotosActivity::class.java)
            intent.putExtra(PlacePhotosActivity.ID, entry.placeId)
            intent.putExtra(PlacePhotosActivity.NAME, entry.placeName)
            startActivity(intent)
        } else {
            Snackbar.make(root, R.string.error_no_internet, Snackbar.LENGTH_SHORT).show()
        }
    }

    companion object {

        val INTENT_KEY = "data"
        val INTENT_ACTION = "action"
        val REQUEST_EDIT_ENTRY = 1

        fun newIntent(context: Context, entry: Entry): Intent {
            val intent = Intent(context, EntryActivity::class.java)
            intent.putExtra(EntryActivity.INTENT_KEY, entry.dateMillis)
            return intent
        }
    }
}
